File: rust/herb-analysis/tests/render_graph_test.rs
use std::collections::BTreeMap;

use herb_analysis::render_graph::{PartialCallSite, RenderGraph, TemplateRoots, Verdict};

fn call_site(caller: &str, ancestors: &[&str], locals: &[&str]) -> PartialCallSite {
  PartialCallSite::render(
    caller,
    locals.iter().map(|local| local.to_string()).collect(),
    ancestors.iter().map(|tag| tag.to_string()).collect(),
  )
}

struct Builder {
  graph: RenderGraph,
}

impl Builder {
  fn new() -> Self {
    Self { graph: RenderGraph::new() }
  }

  fn calls(mut self, partial: &str, sites: Vec<PartialCallSite>) -> Self {
    let mut map = BTreeMap::new();
    let caller = sites.first().expect("at least one call site").caller.clone();

    map.insert(partial.to_string(), sites);
    self.graph.replace_calls_from(&caller, map, 0);

    self
  }

  fn document_root(mut self, file: &str) -> Self {
    self.graph.add_document_root(file);

    self
  }

  fn unresolved(mut self, caller: &str, count: usize) -> Self {
    self.graph.replace_calls_from(caller, BTreeMap::new(), count);

    self
  }

  fn roots(mut self, file: &str, roots: TemplateRoots) -> Self {
    self.graph.set_roots(file, roots);

    self
  }

  fn build(self) -> RenderGraph {
    self.graph
  }
}

fn roots(tags: &[&str], conditional: &[&str], renders: &[&str], resolved: bool) -> TemplateRoots {
  TemplateRoots {
    tags: tags.iter().map(|tag| tag.to_string()).collect(),
    conditional_tags: conditional.iter().map(|tag| tag.to_string()).collect(),
    renders: renders.iter().map(|file| file.to_string()).collect(),
    resolved,
  }
}

#[test]
fn reports_unknown_for_a_partial_nothing_renders() {
  let graph = RenderGraph::new();

  assert_eq!(graph.context_of("_row.html.erb").ancestor_verdict(&[], &["table"]), Verdict::Unknown);
}

#[test]
fn reports_always_when_every_call_site_is_inside_the_tag() {
  let graph = Builder::new()
    .calls("_row.html.erb", vec![call_site("index.html.erb", &["table", "tbody"], &[])])
    .document_root("index.html.erb")
    .build();

  assert_eq!(graph.context_of("_row.html.erb").ancestor_verdict(&[], &["table"]), Verdict::Always);
}

#[test]
fn reports_mixed_when_only_some_call_sites_are_inside_the_tag() {
  let mut graph = RenderGraph::new();
  let mut first = BTreeMap::new();
  first.insert("_row.html.erb".to_string(), vec![call_site("index.html.erb", &["table"], &[])]);
  graph.replace_calls_from("index.html.erb", first, 0);

  let mut second = BTreeMap::new();
  second.insert("_row.html.erb".to_string(), vec![call_site("show.html.erb", &["div"], &[])]);
  graph.replace_calls_from("show.html.erb", second, 0);

  graph.add_document_root("index.html.erb");
  graph.add_document_root("show.html.erb");

  assert_eq!(graph.context_of("_row.html.erb").ancestor_verdict(&[], &["table"]), Verdict::Mixed);
}

#[test]
fn reports_never_when_no_call_site_is_inside_the_tag() {
  let graph = Builder::new()
    .calls("_row.html.erb", vec![call_site("show.html.erb", &["div"], &[])])
    .document_root("show.html.erb")
    .build();

  assert_eq!(graph.context_of("_row.html.erb").ancestor_verdict(&[], &["table"]), Verdict::Never);
}

#[test]
fn reports_unknown_instead_of_never_when_the_graph_is_incomplete() {
  let graph = Builder::new()
    .calls("_row.html.erb", vec![call_site("index.html.erb", &["div"], &[])])
    .unresolved("index.html.erb", 1)
    .build();

  assert_eq!(graph.context_of("_row.html.erb").ancestor_verdict(&[], &["table"]), Verdict::Unknown);
}

#[test]
fn reports_always_from_a_local_ancestor_without_consulting_the_graph() {
  let graph = RenderGraph::new();
  let locals = vec!["table".to_string()];

  assert_eq!(graph.context_of("_row.html.erb").ancestor_verdict(&locals, &["table"]), Verdict::Always);
}

#[test]
fn carries_ancestors_across_two_render_levels() {
  let graph = Builder::new()
    .calls("_cell.html.erb", vec![call_site("_row.html.erb", &["tr"], &[])])
    .calls("_row.html.erb", vec![call_site("index.html.erb", &["table", "tbody"], &[])])
    .document_root("index.html.erb")
    .build();

  let context = graph.context_of("_cell.html.erb");

  assert_eq!(context.chains.len(), 1);
  assert_eq!(context.chains[0].tags, vec!["table", "tbody", "tr"]);
  assert_eq!(context.ancestor_verdict(&[], &["table"]), Verdict::Always);
}

#[test]
fn does_not_loop_forever_on_a_render_cycle() {
  let graph = Builder::new()
    .calls("a.html.erb", vec![call_site("b.html.erb", &["div"], &[])])
    .calls("b.html.erb", vec![call_site("a.html.erb", &["span"], &[])])
    .build();

  assert!(!graph.context_of("a.html.erb").resolved);
}

#[test]
fn finds_the_innermost_matching_ancestor() {
  let graph = Builder::new()
    .calls("_cell.html.erb", vec![call_site("index.html.erb", &["table", "tbody", "tr"], &[])])
    .document_root("index.html.erb")
    .build();

  assert_eq!(
    graph.context_of("_cell.html.erb").closest_ancestor(&[], &["tr", "table"]),
    Some("tr".to_string())
  );
}

#[test]
fn infers_a_signature_from_the_locals_each_call_site_passes() {
  let mut graph = RenderGraph::new();

  let mut first = BTreeMap::new();
  first.insert("_card.html.erb".to_string(), vec![call_site("a.html.erb", &[], &["title", "body"])]);
  graph.replace_calls_from("a.html.erb", first, 0);

  let mut second = BTreeMap::new();
  second.insert("_card.html.erb".to_string(), vec![call_site("b.html.erb", &[], &["title", "footer"])]);
  graph.replace_calls_from("b.html.erb", second, 0);

  let signature = graph.infer_signature("_card.html.erb");

  assert_eq!(
    signature.locals.iter().map(|local| local.name.as_str()).collect::<Vec<_>>(),
    vec!["body", "footer", "title"]
  );
  assert_eq!(signature.call_site_count, 2);
  assert_eq!(signature.strict_locals_declaration(), "<%# locals: (body: nil, footer: nil, title: nil) %>");
}

#[test]
fn adds_a_keyword_rest_to_the_signature_when_the_graph_is_incomplete() {
  let graph = Builder::new()
    .calls("_card.html.erb", vec![call_site("a.html.erb", &[], &["title"])])
    .unresolved("a.html.erb", 1)
    .build();

  assert!(graph.infer_signature("_card.html.erb").strict_locals_declaration().contains("**"));
}

#[test]
fn counts_occurrences_of_a_repeated_chain() {
  let mut graph = RenderGraph::new();

  let mut cell = BTreeMap::new();
  cell.insert("_cell.html.erb".to_string(), vec![call_site("_row.html.erb", &["tr"], &[])]);
  graph.replace_calls_from("_row.html.erb", cell, 0);

  let mut first = BTreeMap::new();
  first.insert("_row.html.erb".to_string(), vec![call_site("index.html.erb", &["table"], &[])]);
  graph.replace_calls_from("index.html.erb", first, 0);

  let mut second = BTreeMap::new();
  second.insert("_row.html.erb".to_string(), vec![call_site("archive.html.erb", &["table"], &[])]);
  graph.replace_calls_from("archive.html.erb", second, 0);

  graph.add_document_root("index.html.erb");
  graph.add_document_root("archive.html.erb");

  let context = graph.context_of("_cell.html.erb");

  assert_eq!(context.chains.len(), 1);
  assert_eq!(context.chains[0].occurrences, 2);
}

#[test]
fn reports_always_for_a_descendant_tag_in_the_partials_own_roots() {
  let graph = Builder::new().roots("_row.html.erb", roots(&["td"], &[], &[], true)).build();

  assert_eq!(graph.descendant_verdict(&["_row.html.erb".to_string()], &["td"]), Verdict::Always);
}

#[test]
fn reports_mixed_for_a_descendant_tag_behind_a_conditional() {
  let graph = Builder::new().roots("_row.html.erb", roots(&[], &["td"], &[], true)).build();

  assert_eq!(graph.descendant_verdict(&["_row.html.erb".to_string()], &["td"]), Verdict::Mixed);
}

#[test]
fn follows_renders_when_deciding_a_descendant_verdict() {
  let graph = Builder::new()
    .roots("_row.html.erb", roots(&[], &[], &["_cell.html.erb"], true))
    .roots("_cell.html.erb", roots(&["td"], &[], &[], true))
    .build();

  assert_eq!(graph.descendant_verdict(&["_row.html.erb".to_string()], &["td"]), Verdict::Always);
}

#[test]
fn reports_unknown_for_a_descendant_when_the_roots_are_unresolved() {
  let graph = Builder::new().roots("_row.html.erb", roots(&[], &[], &[], false)).build();

  assert_eq!(graph.descendant_verdict(&["_row.html.erb".to_string()], &["td"]), Verdict::Unknown);
}

File: test/analysis/render_graph_test.rb
# frozen_string_literal: true

require_relative "../test_helper"
require_relative "../../lib/herb/analysis/render_graph"

class RenderGraphTest < Minitest::Spec
  def call_site(caller, ancestors: [], locals: [], attributes: nil)
    Herb::Analysis::RenderGraph::PartialCallSite.new(
      caller: caller,
      locals: locals,
      ancestors: ancestors,
      ancestor_attributes: attributes,
      via: "render",
      location: nil
    )
  end

  def graph_with(call_sites, document_roots: [], unresolved: {}, roots: {})
    Herb::Analysis::RenderGraph.new(call_sites, roots, Set.new(document_roots), unresolved, Set.new)
  end

  test "reports unknown for a partial nothing renders" do
    graph = graph_with({})

    assert_equal :unknown, graph.context_of("_row.html.erb").ancestor_verdict([], "table")
  end

  test "reports always when every call site is inside the tag" do
    graph = graph_with(
      { "_row.html.erb" => [call_site("index.html.erb", ancestors: ["table", "tbody"])] },
      document_roots: ["index.html.erb"]
    )

    assert_equal :always, graph.context_of("_row.html.erb").ancestor_verdict([], "table")
  end

  test "reports mixed when only some call sites are inside the tag" do
    graph = graph_with(
      {
        "_row.html.erb" => [
          call_site("index.html.erb", ancestors: ["table"]),
          call_site("show.html.erb", ancestors: ["div"])
        ],
      },
      document_roots: ["index.html.erb", "show.html.erb"]
    )

    assert_equal :mixed, graph.context_of("_row.html.erb").ancestor_verdict([], "table")
  end

  test "reports never when no call site is inside the tag" do
    graph = graph_with(
      { "_row.html.erb" => [call_site("show.html.erb", ancestors: ["div"])] },
      document_roots: ["show.html.erb"]
    )

    assert_equal :never, graph.context_of("_row.html.erb").ancestor_verdict([], "table")
  end

  test "reports unknown instead of never when the graph is incomplete" do
    graph = graph_with(
      { "_row.html.erb" => [call_site("index.html.erb", ancestors: ["div"])] },
      unresolved: { "index.html.erb" => 1 }
    )

    assert_equal :unknown, graph.context_of("_row.html.erb").ancestor_verdict([], "table")
  end

  test "reports always from a local ancestor without consulting the graph" do
    graph = graph_with({})

    assert_equal :always, graph.context_of("_row.html.erb").ancestor_verdict(["table"], "table")
  end

  test "carries ancestors across two render levels" do
    graph = graph_with(
      {
        "_cell.html.erb" => [call_site("_row.html.erb", ancestors: ["tr"])],
        "_row.html.erb" => [call_site("index.html.erb", ancestors: ["table", "tbody"])],
      },
      document_roots: ["index.html.erb"]
    )

    context = graph.context_of("_cell.html.erb")

    assert_equal [["table", "tbody", "tr"]], context.chains.map(&:tags)
    assert_equal :always, context.ancestor_verdict([], "table")
  end

  test "does not loop forever on a render cycle" do
    graph = graph_with(
      {
        "a.html.erb" => [call_site("b.html.erb", ancestors: ["div"])],
        "b.html.erb" => [call_site("a.html.erb", ancestors: ["span"])],
      }
    )

    refute graph.context_of("a.html.erb").resolved
  end

  test "finds the innermost matching ancestor" do
    graph = graph_with(
      { "_cell.html.erb" => [call_site("index.html.erb", ancestors: ["table", "tbody", "tr"])] },
      document_roots: ["index.html.erb"]
    )

    assert_equal "tr", graph.context_of("_cell.html.erb").closest_ancestor([], "tr", "table")
  end

  test "infers a signature from the locals each call site passes" do
    graph = graph_with(
      {
        "_card.html.erb" => [
          call_site("a.html.erb", locals: ["title", "body"]),
          call_site("b.html.erb", locals: ["title", "footer"])
        ],
      },
      document_roots: ["a.html.erb", "b.html.erb"]
    )

    signature = graph.infer_signature("_card.html.erb")

    assert_equal ["body", "footer", "title"], signature.locals.map(&:name)
    assert_equal 2, signature.call_site_count
    assert_equal "<%# locals: (body: nil, footer: nil, title: nil) %>", signature.strict_locals_declaration
  end

  test "adds a keyword rest to the signature when the graph is incomplete" do
    graph = graph_with(
      { "_card.html.erb" => [call_site("a.html.erb", locals: ["title"])] },
      unresolved: { "a.html.erb" => 1 }
    )

    assert_equal "<%# locals: (title: nil, **) %>", graph.infer_signature("_card.html.erb").strict_locals_declaration
  end

  test "counts occurrences of a repeated chain" do
    graph = graph_with(
      {
        "_cell.html.erb" => [call_site("_row.html.erb", ancestors: ["tr"])],
        "_row.html.erb" => [
          call_site("index.html.erb", ancestors: ["table"]),
          call_site("archive.html.erb", ancestors: ["table"])
        ],
      },
      document_roots: ["index.html.erb", "archive.html.erb"]
    )

    context = graph.context_of("_cell.html.erb")

    assert_equal 1, context.chains.size
    assert_equal 2, context.chains.first.occurrences
  end

  test "reports always for a descendant tag in the partial's own roots" do
    roots = { "_row.html.erb" => Herb::Analysis::RenderGraph::TemplateRoots.new(tags: ["td"], conditional_tags: [], renders: [], resolved: true) }
    graph = graph_with({}, roots: roots)

    assert_equal :always, graph.descendant_verdict(["_row.html.erb"], "td")
  end

  test "reports mixed for a descendant tag behind a conditional" do
    roots = { "_row.html.erb" => Herb::Analysis::RenderGraph::TemplateRoots.new(tags: [], conditional_tags: ["td"], renders: [], resolved: true) }
    graph = graph_with({}, roots: roots)

    assert_equal :mixed, graph.descendant_verdict(["_row.html.erb"], "td")
  end

  test "follows renders when deciding a descendant verdict" do
    roots = {
      "_row.html.erb" => Herb::Analysis::RenderGraph::TemplateRoots.new(tags: [], conditional_tags: [], renders: ["_cell.html.erb"], resolved: true),
      "_cell.html.erb" => Herb::Analysis::RenderGraph::TemplateRoots.new(tags: ["td"], conditional_tags: [], renders: [], resolved: true),
    }
    graph = graph_with({}, roots: roots)

    assert_equal :always, graph.descendant_verdict(["_row.html.erb"], "td")
  end

  test "round trips through a serialized form" do
    graph = graph_with(
      { "_row.html.erb" => [call_site("index.html.erb", ancestors: ["table"], locals: ["post"])] },
      document_roots: ["index.html.erb"]
    )

    restored = Herb::Analysis::RenderGraph.from(graph.to_h)

    assert_equal graph.to_h, restored.to_h
    assert_equal :always, restored.context_of("_row.html.erb").ancestor_verdict([], "table")
  end
end
